src/buildblock/ML_norm.cxx

```diff
@@ -1336,47 +1481,73 @@ void make_fan_sum_data(Array<2,float>& data_fan_sums, const FanProjData& fan_dat
       data_fan_sums[ra][a] = fan_data.sum(ra,a);
 }
 
-void make_fan_sum_data(Array<2,float>& data_fan_sums,
-		       const ProjData& proj_data)
-{
-  int num_rings;
-  int num_detectors_per_ring;
-  int fan_size;
-  int max_delta;
-  shared_ptr<const ProjDataInfoCylindricalNoArcCorr> proj_data_info_ptr =
-    get_fan_info(num_rings, num_detectors_per_ring, max_delta, fan_size, 
-		 *proj_data.get_proj_data_info_sptr());
+
+template <class TProjDataInfo>
+void make_fan_sum_data_help(Array<2,float>& data_fan_sums,
+                            int num_rings, int num_detectors_per_ring,
+                                     int max_ring_diff, int fan_size,
+                            const TProjDataInfo& proj_data_info,
+                            const ProjData& proj_data)
+{
+
   const int half_fan_size = fan_size/2;
   data_fan_sums.fill(0);
 
-  shared_ptr<SegmentBySinogram<float> > segment_ptr;      
+  shared_ptr<SegmentBySinogram<float> > segment_ptr;
   Bin bin;
 
   for (bin.segment_num() = proj_data.get_min_segment_num(); bin.segment_num() <= proj_data.get_max_segment_num();  ++ bin.segment_num())
   {
     segment_ptr.reset(new SegmentBySinogram<float>(proj_data.get_segment_by_sinogram(bin.segment_num())));
-    
+
     for (bin.axial_pos_num() = proj_data.get_min_axial_pos_num(bin.segment_num());
-	 bin.axial_pos_num() <= proj_data.get_max_axial_pos_num(bin.segment_num());
-	 ++bin.axial_pos_num())
+     bin.axial_pos_num() <= proj_data.get_max_axial_pos_num(bin.segment_num());
+     ++bin.axial_pos_num())
        for (bin.view_num() = 0; bin.view_num() < num_detectors_per_ring/2; bin.view_num()++)
           for (bin.tangential_pos_num() = -half_fan_size;
-	       bin.tangential_pos_num() <= half_fan_size;
+           bin.tangential_pos_num() <= half_fan_size;
                ++bin.tangential_pos_num())
           {
             int ra = 0, a = 0;
             int rb = 0, b = 0;
-            
-            proj_data_info_ptr->get_det_pair_for_bin(a, ra, b, rb, bin);
 
-	    const float value =            
+                proj_data_info.get_det_pair_for_bin(a, ra, b, rb, bin);
+
+        const float value =
               (*segment_ptr)[bin.axial_pos_num()][bin.view_num()][bin.tangential_pos_num()];
-	    data_fan_sums[ra][a] += value;
-	    data_fan_sums[rb][b] += value;
+        data_fan_sums[ra][a] += value;
+        data_fan_sums[rb][b] += value;
           }
   }
 }
 
+void make_fan_sum_data(Array<2,float>& data_fan_sums,
+               const ProjData& proj_data)
+{
+
+    int num_rings;
+    int num_detectors_per_ring;
+    int fan_size;
+    int max_delta;
+    const ProjDataInfo& proj_data_info = *(get_fan_info(num_rings, num_detectors_per_ring, max_delta, fan_size,
```

Comment:
very dangerous. a reference to effectively a returned shared_ptr. This could just as well be cleaned up immediately, and therefore the object pointed to could disappear.

Reply:
so wat is the safe way to pass those info?

Reply:
would have to be
```c++
auto proj_data_info_sptr = get_fan_info(...);
```
However, as I write below, I don't think we need the return value at all.